File: 98.Validate-Binary-Search-Tree.go
// https://leetcode.com/problems/validate-binary-search-tree/description/
//
// algorithms
// Medium (23.75%)
// Total Accepted:    308.7K
// Total Submissions: 1.3M
// beats 100.0% of golang submissions

package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var lastVal int
var res bool
var isFirstEleFound bool

func isValidBST(root *TreeNode) bool {
	if root == nil || (root.Left == nil && root.Right == nil) {
		return true
	}

	res = true
	isFirstEleFound = false

	resursive(root)

	return res
}

func resursive(node *TreeNode) {
	if !res || node == nil {
		return
	}

	resursive(node.Left)

	if !isFirstEleFound {
		lastVal = node.Val
		isFirstEleFound = true
	} else if node.Val > lastVal {
		lastVal = node.Val
	} else {
		res = false
		return
	}

	resursive(node.Right)
}

File: 98.Validate-Binary-Search-Tree.py
# https://leetcode.com/problems/validate-binary-search-tree/
#
# algorithms
# Medium (26.13%)
# Total Accepted:    448,754
# Total Submissions: 1,717,423


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        
        res = [True]
        
        def recursive(node, lower, upper):
            if not node or not res[0]:
                return
            
            if node.val <= lower or node.val >= upper:
                res[0] = False
                return
            
            recursive(node.left, lower, node.val)
            if res[0]:
                recursive(node.right, node.val, upper)
        
        recursive(root, float('-inf'), float('inf'))
        
        return res[0]
